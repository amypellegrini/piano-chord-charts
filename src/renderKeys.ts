import { KeyboardChartLayout, BlackKey, WhiteKey } from "./types";

type OffsetKey = "C" | "D" | "F" | "G" | "A";

type OffsetKeyMap = {
  [key in OffsetKey]: number;
};

const keyNames = ["C", "D", "E", "F", "G", "A", "B"];
const whiteKeyWidth = 23;

const offsetFromWhiteKeyMap: OffsetKeyMap = {
  C: 14.33333,
  D: 18.66666,
  F: 13.25,
  G: 16.25,
  A: 19.75,
};

const flatToSharpMap: { [key: string]: string } = {
  Db: "C#",
  Eb: "D#",
  Gb: "F#",
  Ab: "G#",
  Bb: "A#",
};

const layoutSettingsMap = {
  exact: {
    whiteKeyHeight: 120,
    blackKeyHeight: 80,
  },
  compact: {
    whiteKeyHeight: 65,
    blackKeyHeight: 40,
  },
};

function renderKeys(
  layout: KeyboardChartLayout,
  highlightKeys: Array<WhiteKey | BlackKey>,
  size: number
) {
  const whiteKeyHeight = layoutSettingsMap[layout].whiteKeyHeight;
  const blackKeyHeight = layoutSettingsMap[layout].blackKeyHeight;

  const highlightColour = "#a0c6e8";

  let whiteKeys = "";
  let blackKeys = "";
  let highlightIdx = 0;

  for (let count = 0; count < size; count++) {
    const keyNameIdx = count < 7 ? count : count % 7;
    const keyName = keyNames[keyNameIdx];
    const whiteKeyX = count * whiteKeyWidth;

    let whiteKeyColour = "#fafafa";

    if (highlightKeys) {
      if (
        highlightIdx < highlightKeys.length &&
        highlightKeys[highlightIdx] === keyName
      ) {
        whiteKeyColour = highlightColour;
        highlightIdx += 1;
      }
    }

    whiteKeys = whiteKeys.concat(
      `<rect style="fill:${whiteKeyColour};stroke:black" x="${whiteKeyX}" y="0" width="23" height="${whiteKeyHeight}" ry="3"></rect>\n`
    );

    if (offsetFromWhiteKeyMap[keyName as OffsetKey]) {
      const blackKeyName = keyName + "#";
      const offset = offsetFromWhiteKeyMap[keyName as OffsetKey];
      const blackKeyX = count * whiteKeyWidth + offset;

      let highlightKeyName: WhiteKey | BlackKey =
        (highlightKeys && highlightKeys[highlightIdx]) || "";

      let blackKeycolour = "#222222";

      if (highlightKeyName.match(/b/)) {
        highlightKeyName = flatToSharpMap[highlightKeyName] as
          | WhiteKey
          | BlackKey;
      }

      if (highlightKeys && blackKeyName === highlightKeyName) {
        blackKeycolour = highlightColour;
        highlightIdx++;
      }

      blackKeys = blackKeys.concat(
        `<rect style="fill:${blackKeycolour};stroke:black" x="${blackKeyX}" y="0" width="13" height="${blackKeyHeight}" ry="1"></rect>\n`
      );
    }
  }

  return `${whiteKeys}${blackKeys}`;
}

export default renderKeys;
