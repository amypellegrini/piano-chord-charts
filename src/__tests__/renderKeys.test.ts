import renderKeys from "../renderKeys";
import { BlackKey, KeyboardChartLayout, WhiteKey } from "../types";

describe("renderKeys", () => {
  it.each(["compact", "exact"])(
    "renders keys with the correct layout format",
    (layout) => {
      expect(renderKeys(layout as KeyboardChartLayout, [])).toMatchSnapshot();
    }
  );

  it("highlights specific white keys from left to right", () => {
    const highlightKeys = ["C", "E", "G", "C"];
    const result = renderKeys(
      "compact",
      highlightKeys as Array<WhiteKey | BlackKey>
    );
    expect(result).toMatchSnapshot();
  });

  it("highlights sharp black keys from left to right", () => {
    const highlightKeys: BlackKey[] = ["C#", "D#", "F#"];
    expect(renderKeys("compact", highlightKeys)).toMatchSnapshot();
  });

  it("highlights flat black keys from left to right", () => {
    const highlightKeys: BlackKey[] = ["Db", "Eb", "Gb"];
    const result = renderKeys("compact", highlightKeys);
    expect(result).toMatchSnapshot();
  });
});
