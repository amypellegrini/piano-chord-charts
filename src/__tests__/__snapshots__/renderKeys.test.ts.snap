// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderKeys highlights flat black keys from left to right 1`] = `
"<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"0\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"23\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"46\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"69\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"92\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"115\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"138\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"161\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"184\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"207\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"230\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"253\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"276\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"299\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"14.33333\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"41.66666\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"82.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"108.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"134.75\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"175.33333\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"202.66666\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"243.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"269.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"295.75\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
"
`;

exports[`renderKeys highlights sharp black keys from left to right 1`] = `
"<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"0\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"23\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"46\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"69\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"92\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"115\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"138\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"161\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"184\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"207\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"230\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"253\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"276\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"299\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"14.33333\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"41.66666\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"82.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"108.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"134.75\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"175.33333\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"202.66666\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"243.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"269.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"295.75\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
"
`;

exports[`renderKeys highlights specific white keys from left to right 1`] = `
"<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"0\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"23\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"46\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"69\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"92\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"115\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"138\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#a0c6e8;stroke:black\\" x=\\"161\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"184\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"207\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"230\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"253\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"276\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"299\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"14.33333\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"41.66666\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"82.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"108.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"134.75\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"175.33333\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"202.66666\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"243.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"269.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"295.75\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
"
`;

exports[`renderKeys renders keys with the correct layout format 1`] = `
"<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"0\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"23\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"46\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"69\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"92\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"115\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"138\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"161\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"184\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"207\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"230\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"253\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"276\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"299\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"14.33333\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"41.66666\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"82.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"108.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"134.75\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"175.33333\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"202.66666\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"243.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"269.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"295.75\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
"
`;

exports[`renderKeys renders keys with the correct layout format 2`] = `
"<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"0\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"23\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"46\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"69\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"92\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"115\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"138\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"161\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"184\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"207\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"230\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"253\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"276\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"299\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"14.33333\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"41.66666\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"82.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"108.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"134.75\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"175.33333\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"202.66666\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"243.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"269.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"295.75\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
"
`;

exports[`renderKeys supports custom size 1`] = `
"<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"0\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"23\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"46\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"69\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"92\\" y=\\"0\\" width=\\"23\\" height=\\"65\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"14.33333\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"41.66666\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"82.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
<rect style=\\"fill:#222222;stroke:black\\" x=\\"108.25\\" y=\\"0\\" width=\\"13\\" height=\\"40\\" ry=\\"1\\"></rect>
"
`;
